File: src/runtime/value/value_function.cpp
#include "value_function.h"
#include "runtime/space.h"
#include "runtime/variable.h"
#include "value_tool.h"

using namespace peak;

ValueFunction::ValueFunction(std::size_t paramSize, FunctionType func)
	: _params(paramSize), _function(func) {
	for (auto i = 0u; i < paramSize; ++i) {
		_params[i] = std::string("%") + std::to_string(i);
	}
}
ValueFunction::ValueFunction(const std::vector<std::string>& params, FunctionType func)
	: _params(params), _function(func) {
}

ValueFunction::ValueFunction() {
}

std::shared_ptr<Value> ValueFunction::Call(const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space> space) {
	if (args.size() != _params.size()) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::FunctionCall, "The function (" + std::to_string(args.size()) + " params) not found!");
		return nullptr;
	}

	auto tempSpace = std::make_shared<Space>(SpaceType::Function, space);
	for (std::size_t i = 0; i < args.size(); ++i) {
		auto tempVariable = std::make_shared<Variable>(_params[i], VariableAttribute::None);
		if (!tempSpace->AddVariable(tempVariable)) {
			ErrorLogger::LogRuntimeError(ErrorRuntimeCode::FunctionCall, "The function params name is exist!");
			return nullptr;
		}
		tempVariable->SetValue(args[i]);
	}
	auto result = _function(args, tempSpace);
	tempSpace->Clear();
	return result;
}

std::string ValueFunction::ToString() const {
	std::string ret = "function (";
	for (auto i = 0u; i < _params.size(); ++i) {
		ret += _params[i];
		if (i < _params.size() - 1) {
			ret += ", ";
		}
	}
	ret += ")";
	return ret;
}
std::shared_ptr<Value> ValueFunction::Clone() const {
	return std::make_unique<ValueFunction>(_params, _function);

}

File: src/Runtime/Value/ValueFunction.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System;
using System.Collections.Generic;

namespace peak.interpreter {

	public class ValueFunction : TypeValue<ValueFunction> {

		private List<string> _params = new List<string>();
		private Func<List<Value>, Space, Value> _function = null;
		private ValueFunction() {}
		public ValueFunction(int paramSize, Func<List<Value>, Space, Value> func) {
			for (var i=0; i<paramSize; ++i)
			{
				_params.Add("%" + i);
			}
			_function = func;
		}
		public ValueFunction(List<string> params_, Func<List<Value>, Space, Value> func) {
			_params = params_;
			_function = func;
		}
		public Value Call(List<Value> args, Space space) {
			if (_params.Count != args.Count) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.FunctionCall, "The function (" + args.Count + " params) not found!");
				return null;
			}
			var tempSpace = new Space(SpaceType.Function, space);
			for (int i = 0; i < args.Count; ++i) {
				var tempVariable = new Variable(_params[i], VariableAttribute.None);
				if (!tempSpace.AddVariable(tempVariable)) {
					ErrorLogger.LogRuntimeError(ErrorRuntimeCode.FunctionCall, "The function params name is exist!");
					return null;
				}
				tempVariable.SetValue(args[i]);
			}
			var result = _function.Invoke(args, tempSpace);
			tempSpace.Clear();
			return result;
		}
		public override string ToString() {
			string ret = "function (";
			for (var i = 0; i < _params.Count; ++i)
			{
				ret += _params[i];
				if (i < _params.Count - 1)
				{
					ret += ", ";
				}
			}
			ret += ")";
			return ret;
		}
		public override Value Clone() {
			return new ValueFunction(_params, _function);
		}
	}
} // namespace peak.interpreter